packages/mobile/src/web3/contracts.ts

```diff
@@ -48,20 +63,58 @@ const getWeb3IpcProvider = (testnet: Testnets) => {
   return ipcProvider
 }
 
-const getWeb3HttpProvider = (testnet: Testnets) => {
-  Logger.debug(tag, 'creating HttpProvider...')
+// Use Http provider on iOS until we add support for local socket on iOS in react-native-tcp
+function getWeb3HttpProviderForIos(): Provider {
+  Logger.debug(tag, 'creating HttpProvider for iOS...')
 
   const httpProvider = new Web3.providers.HttpProvider('http://localhost:8545')
-  Logger.debug(tag, 'created HttpProvider')
+  Logger.debug(tag, 'created HttpProvider for iOS')
 
   return httpProvider
 }
 
-// Use Http provider on iOS until we add support for local socket on iOS in react-native-tcp
-export const getWeb3Provider = Platform.OS === 'ios' ? getWeb3HttpProvider : getWeb3IpcProvider
+function getWebSocketProvider (url: string): Provider {
+  Logger.debug(tag, 'creating HttpProvider...')
+  const provider = new Web3.providers.HttpProvider(url)
+  Logger.debug(tag, 'created HttpProvider')
+  // In the future, we might decide to over-ride the error handler via the following code.
+  // provider.on('error', () => {
+  //   Logger.showError('Error occurred')
+  // })
+  return provider
+}
+
+function getWeb3(): Web3 {
+  Logger.info(`Initializing web3, platform: ${Platform.OS}, geth free mode: ${isZeroSyncMode()}`)
 
-export const setWeb3Provider = (testnet: Testnets) => {
-  web3.setProvider(getWeb3Provider(testnet))
+  if (isZeroSyncMode() && (Platform.OS === 'ios')) {
+        throw new Error('Zero sync mode is currently not supported on iOS')
+  } else if (isZeroSyncMode()) { // Geth free mode
+    const url = getInfuraUrl()
+    Logger.debug('contracts@getWeb3', `Connecting to url ${url}`)
+    return new Web3(getWebSocketProvider(url))
+  } else if (Platform.OS === 'ios') {  // iOS + local geth
+    return new Web3(getWeb3HttpProviderForIos())
+  } else {
+    return new Web3(getIpcProvider(DEFAULT_TESTNET))
+  }
 }
 
-export let web3 = new Web3(getWeb3Provider(DEFAULT_TESTNET))
+export function addLocalAccount(web3Instance: Web3, privateKey: string): Web3 {
+  if (!isZeroSyncMode()) {
+    throw new Error('addLocalAccount can only be called in Zero sync mode')
+  }
+  if (web3Instance === null) {
+    throw new Error('web3 instance is null')
+  }
+  if (web3Instance === undefined) {
```

Comment:
Definitely appreciate the thoroughness but `if (!web3Instance)` is more inline with our other invariant checks and what's typical in javascript :)

Reply:
Changed to
```
if (!web3Instance) {
    throw new Error(`web3 instance is ${web3Instance}`)
  }
```